examples/skip.json

```diff
@@ -0,0 +1,5 @@
+{
```

Comment:
I'm not sure if we actually need such a detailed filters for now. Initially I was thinking of the following requirements:
* being able to do both: include **and** exclude tests (e.g. `--filter-include ".*thread.*" --filter-exclude ".*open.*"`)
* pass regular expressions so we can e.g. disable the whole testsuite (something like `--filter-exclude "WASI C tests.*"`)

But I'm curious what are your thoughts on that.

Reply:
my intention is to implement something simplest first. (regex etc is more complex and advanced in my taste.)

also i think it makes sense to have a filter in a file because i guess it's almost static for a given runtime.

Reply:
Sure, I don't suggest implementing all that at once, but I'd like us to think a bit about the API before we move forward. The proposed API of the JSON file only allows for disabling tests, and I'm afraid that if we want to add more features (like the ones mentioned above) we'll either have to make backward-incompatible changes or make the API inconsistent. As I said, we don't have to implement everything in the first iteration, but at least having placeholders would probably simplify further development.

Reply:
i feel compatibility concern is not too important at this point.

also, your requirements are not clear to me.
if you explain it a bit more, maybe i can suggest a schema.

Reply:
Sure; so first of all, I think we should be able to allow users to both select exclusion and inclusion filter (at the moment only the exclusion is possible). Also, we might allow for wildcard/regex, but I assume this could be just a feature request for the future, and we can implement it in the future; just an idea (although feel free to come up with something different):
```json
{
  "include": [], // if empty, all tests are included
  "exclude/ignore(whatever sounds better)": [] // if empty, none of the tests are excluded
}
```
Element in the array could be a full path to a test, e.g. `WASI C tests.clock_getres-monotonic`, so we can in the future implement wildcards to only enable clock_getres tests, with: `WASI C tests.clock_getres*`.

As I said, this is just an example schema to explain what I mean; please let me know if it's clear what we try to achieve here and share your ideas.

Reply:
> Sure; so first of all, I think we should be able to allow users to both select exclusion and inclusion filter (at the moment only the exclusion is possible). Also, we might allow for wildcard/regex, but I assume this could be just a feature request for the future, and we can implement it in the future; just an idea (although feel free to come up with something different):
> 
> ```json
> {
>   "include": [], // if empty, all tests are included
>   "exclude/ignore(whatever sounds better)": [] // if empty, none of the tests are excluded
> }
> ```

if a test is included in both of "include" and "exclude", "exclude" wins, right?

> 
> Element in the array could be a full path to a test, e.g. `WASI C tests.clock_getres-monotonic`, so we can in the future implement wildcards to only enable clock_getres tests, with: `WASI C tests.clock_getres*`.

while regex might be convenient for some cases, to enable/disable a set of tests,
it might be simpler to have "feature" flags in each tests.
eg. "require": ["wasi-threads"]

> 
> As I said, this is just an example schema to explain what I mean; please let me know if it's clear what we try to achieve here and share your ideas.

Reply:
> if a test is included in both of "include" and "exclude", "exclude" wins, right?

I don't have preference about that; maybe worth checking what other frameworks do for consistency?

> while regex might be convenient for some cases, to enable/disable a set of tests,
> it might be simpler to have "feature" flags in each tests.
> eg. "require": ["wasi-threads"]
I think we can have both; I also like the idea of having `feature`/`tags` flags; in fact, I had similar idea to track what wasi functions are tested in a given test (see https://github.com/WebAssembly/wasi-testsuite/blob/6d4b57497ceacd304e35a7ab03dd9bb65950bd79/test-runner/wasi_test_runner/test_case.py#L36), maybe we can re-purpose this field to be a `feature`/`tag` field?

Reply:
my suggestion for this PR:

* rename `--filter` option to `--exclude-filter`
* no changes to schema
* we can consider adding `--include-filter`, `--exclude-regex-filter` etc later when/if necessary.

Reply:
> rename `--filter` option to `--exclude-filter`
Given we already define filters in the JSON file, I don't know if there's a need for having two cli parameters, or instead could we have just one (`--filter`) and update the schema so it allows for specifying both inclusion and exclusion - I don't have a strong opinion; just a thought, as I haven't seen them separate in other farameworks. 

> no changes to schema
Not sure if we actually need a nested structure here (testsuite->tests); if we implement regex/glob we could probably just have a list of strings? We can start with this one as it might seem a bit simpler but once we update it, we'll have to update all the consumers as well (not sure how many runtimes will onboard by then though).

I'm also not 100% sure about file vs cli parameter. Whereas I see the benefit you mentioned above, I think it 
doesn't matter that much for runtimes as they'll have their own CI scripts that wrap around the call to the tests. The benefit of having everything through CLI is that developers can easily just filter some of the tests they're currently working on without modifying file (and risking commiting it accidentally). Similar to my previous point, it is something we can change in the future, but depending on the number of adopters, it might be easier or harder.

Reply:
> > rename `--filter` option to `--exclude-filter`
> 
> Given we already define filters in the JSON file, I don't know if there's a need for having two cli parameters, or instead could we have just one (`--filter`) and update the schema so it allows for specifying both inclusion and exclusion - I don't have a strong opinion; just a thought, as I haven't seen them separate in other farameworks.

given the current cli options which allows to specify multiple testsuites,
i thought it's natural to accept multiple filters.

> 
> > no changes to schema
> 
> Not sure if we actually need a nested structure here (testsuite->tests); if we implement regex/glob we could probably just have a list of strings? We can start with this one as it might seem a bit simpler but once we update it, we'll have to update all the consumers as well (not sure how many runtimes will onboard by then though).

i'm not the person who invented the testsuite->tests structure in this repo.
honestly speaking i don't think the testsuite concept makes much sense.

maybe we can use some kind of string match with test id, where test id is:
```
test_id = testsuite + '/' + test
```

> 
> I'm also not 100% sure about file vs cli parameter. Whereas I see the benefit you mentioned above, I think it doesn't matter that much for runtimes as they'll have their own CI scripts that wrap around the call to the tests. The benefit of having everything through CLI is that developers can easily just filter some of the tests they're currently working on without modifying file (and risking commiting it accidentally). Similar to my previous point, it is something we can change in the future, but depending on the number of adopters, it might be easier or harder.

as a developer, i feel it easier to copy/modify a file than tweaking cli options.
but i can understand you have a different preference.

Reply:
To unblock the change, let's move on with your suggestion here https://github.com/WebAssembly/wasi-testsuite/pull/35#discussion_r1063327236, we'll adapt it once we have more feedback.

Reply:
> To unblock the change, let's move on with your suggestion here [#35 (comment)](https://github.com/WebAssembly/wasi-testsuite/pull/35#discussion_r1063327236), we'll adapt it once we have more feedback.

done